Index: skeleton/remote/protocol/XpProtocolHandler.class.php
===================================================================
--- skeleton/remote/protocol/XpProtocolHandler.class.php	(revision 7678)
+++ skeleton/remote/protocol/XpProtocolHandler.class.php	(working copy)
@@ -23,7 +23,8 @@
     var
       $versionMajor   = 0,
       $versionMinor   = 0,
-      $serializer     = NULL;
+      $serializer     = NULL,
+      $cat            = NULL;
     
     var
       $_sock= NULL;  
@@ -44,6 +45,26 @@
       $this->serializer->exceptionName('invoke/Exception', 'remote.InvocationException');
       $this->serializer->packageMapping('net.xp_framework.easc.reflect', 'remote.reflect');
     }
+    
+    /**
+     * Create a string representation of a given value
+     *
+     * @access  protected
+     * @param   &mixed value
+     * @return  string
+     */
+    function stringOf(&$value) {
+      if (is_a($value, 'Proxy')) {
+        $s= 'Proxy<';
+        $c= get_class($value);
+        $implements= xp::registry('implements');
+        foreach (array_keys($implements[$c]) as $iface) {
+          $s.= xp::nameOf($iface).', ';
+        }
+        return substr($s, 0, -2).'>';
+      }
+      return xp::stringOf($value);
+    }
 
     /**
      * Initialize this protocol handler
@@ -63,13 +84,27 @@
       $this->_sock->connect();
       
       if ($user= $proxy->getUser()) {
-        $this->sendPacket(REMOTE_MSG_INIT, "\1", array(
+        $this->cat && $this->cat->infof(
+          '>>> %s(%s:%d) INITIALIZE %s',
+          $this->getClassName(),
+          $this->_sock->host,
+          $this->_sock->port,
+          $user
+        );
+        $r= $this->sendPacket(REMOTE_MSG_INIT, "\1", array(
           new ByteCountedString($proxy->getUser()),
           new ByteCountedString($proxy->getPassword())
         ));
       } else {
-        $this->sendPacket(REMOTE_MSG_INIT, "\0");
+        $this->cat && $this->cat->infof(
+          '>>> %s(%s:%d) INITIALIZE',
+          $this->getClassName(),
+          $this->_sock->host,
+          $this->_sock->port
+        );
+        $r= $this->sendPacket(REMOTE_MSG_INIT, "\0");
       }
+      $this->cat && $this->cat->infof('<<< %s', $this->stringOf($r));
     }
     
     /**
@@ -90,7 +125,16 @@
      * @param   &lang.Object
      */
     function &lookup($name) {
-      return $this->sendPacket(REMOTE_MSG_LOOKUP, '', array(new ByteCountedString($name)));
+      $this->cat && $this->cat->infof(
+        '>>> %s(%s:%d) LOOKUP %s',
+        $this->getClassName(),
+        $this->_sock->host,
+        $this->_sock->port,
+        $name
+      );
+      $r= &$this->sendPacket(REMOTE_MSG_LOOKUP, '', array(new ByteCountedString($name)));
+      $this->cat && $this->cat->infof('<<< %s', $this->stringOf($r));
+      return $r;
     }
 
     /**
@@ -101,7 +145,16 @@
      * @param   bool
      */
     function begin(&$tran) {
-      return $this->sendPacket(REMOTE_MSG_TRAN_OP, pack('N', REMOTE_TRAN_BEGIN));
+      $this->cat && $this->cat->infof(
+        '>>> %s(%s:%d) BEGIN %s',
+        $this->getClassName(),
+        $this->_sock->host,
+        $this->_sock->port,
+        $this->stringOf($tran)
+      );
+      $r= &$this->sendPacket(REMOTE_MSG_TRAN_OP, pack('N', REMOTE_TRAN_BEGIN));
+      $this->cat && $this->cat->infof('<<< %s', $this->stringOf($r));
+      return $r;
     }
 
     /**
@@ -112,7 +165,16 @@
      * @param   bool
      */
     function rollback(&$tran) {
-      return $this->sendPacket(REMOTE_MSG_TRAN_OP, pack('N', REMOTE_TRAN_ROLLBACK));
+      $this->cat && $this->cat->infof(
+        '>>> %s(%s:%d) ROLLBACK %s',
+        $this->getClassName(),
+        $this->_sock->host,
+        $this->_sock->port,
+        $this->stringOf($tran)
+      );
+      $r= &$this->sendPacket(REMOTE_MSG_TRAN_OP, pack('N', REMOTE_TRAN_ROLLBACK));
+      $this->cat && $this->cat->infof('<<< %s', $this->stringOf($r));
+      return $r;
     }
 
     /**
@@ -123,7 +185,16 @@
      * @param   bool
      */
     function commit(&$tran) {
-      return $this->sendPacket(REMOTE_MSG_TRAN_OP, pack('N', REMOTE_TRAN_COMMIT));
+      $this->cat && $this->cat->infof(
+        '>>> %s(%s:%d) COMMIT %s',
+        $this->getClassName(),
+        $this->_sock->host,
+        $this->_sock->port,
+        $this->stringOf($tran)
+      );
+      $r= &$this->sendPacket(REMOTE_MSG_TRAN_OP, pack('N', REMOTE_TRAN_COMMIT));
+      $this->cat && $this->cat->infof('<<< %s', $this->stringOf($r));
+      return $r;
     }
 
     /**
@@ -137,6 +208,15 @@
      * @return  &mixed
      */
     function &invoke($oid, $method, $args) {
+      $this->cat && $this->cat->infof(
+        '>>> %s(%s:%d) %d::%s(%s)',
+        $this->getClassName(),
+        $this->_sock->host,
+        $this->_sock->port,
+        $oid,
+        $method,
+        $this->stringOf($args)
+      );
       $r= &$this->sendPacket(
         REMOTE_MSG_CALL, 
         pack('NN', 0, $oid),
@@ -145,6 +225,7 @@
           new ByteCountedString($this->serializer->representationOf(new ArrayList($args)))
         )
       );
+      $this->cat && $this->cat->infof('<<< %s', $this->stringOf($r));
       return $r;
     }
 
@@ -178,10 +259,10 @@
         $length,
         $data
       );
-      // DEBUG Console::writeLine('>>>', addcslashes($packet, "\0..\37!@\177..\377"));
 
       try(); {
         $this->_sock->write($packet);
+        $this->cat && $this->cat->debug('>>> Request:', $this->stringOf($bytes));
         for ($i= 0; $i < $bsize; $i++) {
           $bytes[$i]->writeTo($this->_sock);
         }
@@ -193,7 +274,6 @@
         return throw(new RemoteException($e->getMessage(), $e));
       }
       
-      // DEBUG Console::writeLine('<<<', xp::stringOf($header));
       if (DEFAULT_PROTOCOL_MAGIC_NUMBER != $header['magic']) {
         $this->_sock->close();
         return throw(new Error('Magic number mismatch (have: '.$header['magic'].' expect: '.DEFAULT_PROTOCOL_MAGIC_NUMBER));
@@ -205,26 +285,30 @@
         switch ($header['type']) {
           case REMOTE_MSG_VALUE:
             $data= &ByteCountedString::readFrom($this->_sock);
-            // Console::writeLine('<<<', addcslashes($data, "\0..\37!@\177..\377"));
+            $this->cat && $this->cat->debug('<<< Response:', addcslashes($data, "\0..\37!@\177..\377"));
             return $this->serializer->valueOf($data, $length= 0, $ctx);
 
           case REMOTE_MSG_EXCEPTION:
-            $reference= &$this->serializer->valueOf(ByteCountedString::readFrom($this->_sock), $length= 0, $ctx);
+            $data= &ByteCountedString::readFrom($this->_sock);
+            $this->cat && $this->cat->debug('<<< Response:', addcslashes($data, "\0..\37!@\177..\377"));
+            $reference= &$this->serializer->valueOf($data, $length= 0, $ctx);
             if (is('RemoteException', $reference)) {
               return throw($reference);
             } else if (is('ClassReference', $reference)) {
               return throw(new RemoteException($reference->getClassName(), $reference));
             } else {
-              return throw(new RemoteException('lang.Exception', new Exception(xp::stringOf($reference))));
+              return throw(new RemoteException('lang.Exception', new Exception($this->stringOf($reference))));
             }
 
           case REMOTE_MSG_ERROR:
             $message= ByteCountedString::readFrom($this->_sock);    // Not serialized!
+            $this->cat && $this->cat->debug('<<< Response:', addcslashes($message, "\0..\37!@\177..\377"));
             $this->_sock->close();
             return throw(new RemoteException($message, new Error($message)));
 
           default:
-            $this->readBytes($header['length']);   // Read all left-over bytes
+            $data= &$this->readBytes($header['length']);   // Read all left-over bytes
+            $this->cat && $this->cat->debug('<<< Response:', addcslashes($data, "\0..\37!@\177..\377"));
             $this->_sock->close();
             return throw(new Error('Unknown message type'));
         }
@@ -249,5 +333,15 @@
       return $return;
     }
 
-  } implements(__FILE__, 'remote.protocol.ProtocolHandler');
+    /**
+     * Set trace
+     *
+     * @accessp public
+     * @param &util.log.LogCategory cat
+     */
+    function setTrace(&$cat) {
+      $this->cat= &$cat;
+    }
+
+  } implements(__FILE__, 'remote.protocol.ProtocolHandler', 'util.log.Traceable');
 ?>
Index: skeleton/remote/Remote.class.php
===================================================================
--- skeleton/remote/Remote.class.php	(revision 7678)
+++ skeleton/remote/Remote.class.php	(working copy)
@@ -5,7 +5,6 @@
  */
 
   uses(
-    'peer.URL', 
     'remote.HandlerInstancePool', 
     'remote.protocol.RemoteInterfaceMapping', 
     'remote.UserTransaction'
@@ -72,12 +71,10 @@
         if (isset($instances[$key])) return $instances[$key];
 
         // No instance yet, so get it
-        $url= &new URL($key);
         $e= $instance= NULL;
         try(); {
           $instance= &new Remote();
-          $instance->_handler= &$pool->acquire($url);
-          $instance->_handler && $instance->_handler->initialize($url);
+          $instance->_handler= &$pool->acquire($key, TRUE);
         } if (catch('RemoteException', $e)) {
           continue;   // try next
         } if (catch('Exception', $e)) {
Index: skeleton/remote/HandlerInstancePool.class.php
===================================================================
--- skeleton/remote/HandlerInstancePool.class.php	(revision 7678)
+++ skeleton/remote/HandlerInstancePool.class.php	(working copy)
@@ -4,7 +4,7 @@
  * $Id$ 
  */
 
-  uses('remote.HandlerFactory', 'util.collections.HashTable');
+  uses('remote.HandlerFactory', 'util.collections.HashTable', 'peer.URL');
 
   /**
    * Pool of handler instances
@@ -12,9 +12,10 @@
    * @see      xp://remote.HandlerFactory
    * @purpose  Pool
    */
-  class HandlerInstancePool extends HashTable {
+  class HandlerInstancePool extends Object {
     var
-      $pool= NULL;
+      $pool = NULL,
+      $cat  = NULL;
 
     /**
      * Constructor
@@ -70,21 +71,26 @@
      * Acquire a handler instance
      *
      * @access  public
-     * @param   &peer.URL url
+     * @param   string key
      * @return  &remote.protocol.ProtocolHandler
      * @throws  remote.protocol.UnknownProtocolException
      */
-    function &acquire(&$url) {
-      if ($this->pool->containsKey($url)) return $this->pool->get($url);
+    function &acquire($key, $initialize= FALSE) {
+      $url= &new URL($key);
+      if ($this->pool->containsKey($url)) {
+        $instance= &$this->pool->get($url);
+      } else {
+        sscanf($url->getScheme(), '%[^+]+%s', $type, $option);
+        try(); {
+          $class= &HandlerFactory::handlerFor($type);
+        } if (catch('Exception', $e)) {
+          return throw($e);
+        }
 
-      sscanf($url->getScheme(), '%[^+]+%s', $type, $option);
-      try(); {
-        $class= &HandlerFactory::handlerFor($type);
-      } if (catch('Exception', $e)) {
-        return throw($e);
+        $instance= &$this->pool($url, $class->newInstance($option));
       }
 
-      $instance= &$this->pool($url, $class->newInstance($option));
+      $initialize && $instance->initialize($url);
       return $instance;
     }
   }
Index: ext/java/easc/clients/echo.php
===================================================================
--- ext/java/easc/clients/echo.php	(revision 7678)
+++ ext/java/easc/clients/echo.php	(working copy)
@@ -9,6 +9,8 @@
     'remote.Remote', 
     'util.cmd.ParamString', 
     'util.profiling.Timer',
+    'util.log.Logger',
+    'util.log.ColoredConsoleAppender',
     'util.profiling.unittest.AssertionFailedError'
   );
   
@@ -20,7 +22,7 @@
 
 Usage
 -----
-$ php echo.php <hostname> <type>  [-p <port> ] [-j <jndi_name> ]
+$ php echo.php <hostname> <type>  [-p <port> ] [-j <jndi_name> ] [-v level]
   
   * hostname is the host name (or IP) that your JBoss + XP-MBean server 
     is running on. The feed entity bean (from the easc/beans directory) 
@@ -40,13 +42,37 @@
   
   * jndi_name is the name of the bean in JNDI. It defaults to 
     "xp/demo/Echo"
+  
+  * Level is the log level, defaulting to all, and may consist of one 
+    or more of debug, info, warn or error (separated by commas), which
+    would selectively activate the specified log level.
 __
     );
     exit(1);
   }
   
+  $url= 'xp://'.$p->value(1).':'.$p->value('port', 'p', 6448).'/';
+  
+  // Debugging
+  $cat= NULL;
+  if ($p->exists('verbose')) {
+    $l= &Logger::getInstance();
+    $cat= &$l->getCategory();
+    $cat->addAppender(new ColoredConsoleAppender());
+    $pool= &HandlerInstancePool::getInstance();
+    $handler= &$pool->acquire($url);
+    $handler->setTrace($cat);
+    
+    if ($levels= $p->value('verbose')) {
+      foreach (explode(',', $levels) as $level) {
+        $flags= $flags | constant('LOGGER_FLAG_'.strtoupper($level));
+      }
+      $cat->setFlags($flags);
+    }
+  }
+  
   try(); {
-    $remote= &Remote::forName('xp://'.$p->value(1).':'.$p->value('port', 'p', 6448).'/');
+    $remote= &Remote::forName($url);
     $remote && $home= &$remote->lookup($p->value('jndi', 'j', 'xp/demo/Roundtrip'));
     $home && $instance= &$home->create();
   } if (catch('Exception', $e)) {
Index: skeleton/net/xp_framework/unittest/remote/RoundtripTest.class.php
===================================================================
--- skeleton/net/xp_framework/unittest/remote/RoundtripTest.class.php	(revision 7678)
+++ skeleton/net/xp_framework/unittest/remote/RoundtripTest.class.php	(working copy)
@@ -159,7 +159,7 @@
      *
      * @access  public
      */
-    #[@test, @expect('remote.RemoteException')]
+    #[@ignore('Method overloading not supported by PHP'), @test, @expect('remote.RemoteException')]
     function incorrectArgumentsToArrayMethod() {
       $this->instance->echoArray('A STRING, MAN!');
     }
Index: skeleton/net/xp_framework/unittest/remote/RemoteTest.class.php
===================================================================
--- skeleton/net/xp_framework/unittest/remote/RemoteTest.class.php	(revision 7678)
+++ skeleton/net/xp_framework/unittest/remote/RemoteTest.class.php	(working copy)
@@ -54,7 +54,7 @@
         REMOTE_SPEC_THREE   => FALSE,   // Cluster machine #3
         REMOTE_SPEC_OTHER   => TRUE     // Other machine
       ) as $spec => $avail) {
-        $this->handler[$spec]= &$pool->acquire(new URL($spec));
+        $this->handler[$spec]= &$pool->acquire($spec);
         $this->handler[$spec]->server['available']= $avail;
       }
     }
