Index: Runner.class.php
===================================================================
--- Runner.class.php	(revision 10852)
+++ Runner.class.php	(working copy)
@@ -178,13 +178,37 @@
       if (!$classname) {
         self::$err->writeLine('*** Missing classname');
         return 1;
+      } else if (strstr($classname, xp::CLASS_FILE_EXT)) {
+        $file= new File($classname);
+        if (!$file->exists()) {
+          self::$err->writeLine('Cannot load class from non-existant file ', $classname);
+          return 1;
+        }
+        $uri= $file->getURI();
+        $path= dirname($uri);
+        $paths= array_flip(array_map('realpath', xp::$registry['classpath']));
+        $class= NULL;
+        while (FALSE !== ($pos= strrpos($path, DIRECTORY_SEPARATOR))) { 
+          if (isset($paths[$path])) {
+            $class= XPClass::forName(strtr(substr($uri, strlen($path)+ 1, -10), DIRECTORY_SEPARATOR, '.'));
+            break;
+          }
+
+          $path= substr($path, 0, $pos); 
+        }
+
+        if (!$class) {
+          self::$err->writeLine('Cannot load class from ', $file);
+          return 1;
+        }
+      } else {
+        try {
+          $class= XPClass::forName($classname);
+        } catch (ClassNotFoundException $e) {
+          self::$err->writeLine('*** ', $e->getMessage());
+          return 1;
+        }
       }
-      try {
-        $class= XPClass::forName($classname);
-      } catch (ClassNotFoundException $e) {
-        self::$err->writeLine('*** ', $e->getMessage());
-        return 1;
-      }
 
       // Usage
       if ($classparams->exists('help', '?')) {
@@ -255,13 +279,13 @@
           try {
             $method->invoke($instance, array($pass));
           } catch (Throwable $e) {
-            self::$err->writeLine('*** Error for argument '.$name.': '.$e->getMessage());
+            self::$err->writeLine('*** Error for arguments '.$begin.'..'.$end.': '.$e->getMessage());
             return 2;
           }
         } else if ($method->hasAnnotation('arg')) {  // Pass arguments
           $arg= $method->getAnnotation('arg');
           if (isset($arg['position'])) {
-            $name= '#'.$arg['position'];
+            $name= '#'.($arg['position']+ 1);
             $select= intval($arg['position']);
             $short= NULL;
           } else if (isset($arg['name'])) {
@@ -293,13 +317,6 @@
             self::$err->writeLine('*** Error for argument '.$name.': '.$e->getMessage());
             return 2;
           }
-        } else if ($method->hasAnnotation('args')) { // Pass all arguments
-          try {
-            $method->invoke($instance, array($classparams->list));
-          } catch (Throwable $e) {
-            self::$err->writeLine('*** Error for argument '.$name.': '.$e->getMessage());
-            return 2;
-          }
         }
       }
 
