Index: lang.base.php
===================================================================
--- lang.base.php	(revision 7907)
+++ lang.base.php	(working copy)
@@ -227,6 +227,15 @@
   // {{{ void uses (string* args)
   //     Uses one or more classes
   function uses() {
+    static $bootstrap= array();
+    static $h;
+    $include= &xp::registry('include_path');
+
+    if (0 == sizeof($include)) {
+      $include= array_flip(explode(':', ini_get('include_path')));
+      xp::registry('include_path', $include);
+    }
+    
     foreach (func_get_args() as $str) {
       if (class_exists($class= xp::reflect($str))) continue;
 
@@ -245,14 +254,75 @@
         }
         $str= substr($str, strrpos($str, '/')+ 1);
         $class= xp::reflect($str);
-      } else {
-        if (FALSE === ($r= include_once(strtr($str, '.', DIRECTORY_SEPARATOR).'.class.php'))) {
-          xp::error(xp::stringOf(new Error('Cannot include '.$str)));
-        } else if (TRUE === $r) {
-          continue;
+      }
+
+      $t1= microtime();
+      foreach ($include as $path => $loader) {
+        $type= 'NONE';
+
+        // If path is a directory and the included file exists, load it
+        if (is_dir($path) && file_exists($f= realpath($path.DIRECTORY_SEPARATOR.strtr($str, '.', DIRECTORY_SEPARATOR).'.class.php'))) {
+          if (FALSE === ($r= include_once($f))) {
+            xp::error(xp::stringOf(new Error('Cannot include '.$str)));
+          }
+          
+          $type= 'DIRLDR';
+          break;
+        } elseif (is_file($path) && class_exists('archiveclassloader') && class_exists('archive') && class_exists('file')) {
+
+          // Load via archive class loader
+          if (!is_object($include[$path])) {
+            $include[$path]= &new ArchiveClassLoader(new Archive(new File($path)));
+            xp::registry('include_path', $include);
+          }
+          
+          if (!$include[$path]->providesClass($str)) continue;
+          $class= &$include[$path]->loadClass($str);
+          
+          $type= 'ARCLDR';
+          break;
+        } elseif (is_file($path) && ('lang.' === substr($str, 0, 5) || 'io.' === substr($str, 0, 3))) {
+
+          // Bootstrap loading, only to be used for core classes.
+          if (!is_resource($h)) {
+            $h= fopen($path, 'rb');
+            $header= unpack('a3id/c1version/i1indexsize/a*reserved', fread($h, 0x0100));
+            for ($bootstrap[$path]= array(), $i= 0; $i < $header['indexsize']; $i++) {
+              $entry= unpack(
+                'a80id/a80filename/a80path/i1size/i1offset/a*reserved', 
+                fread($h, 0x0100)
+              );
+              $bootstrap[$path][$entry['id']]= array($entry['size'], $entry['offset']);
+            }
+          }
+          
+          if (!isset($bootstrap[$path][$str])) continue;
+          fseek($h, 0x0100 + sizeof($bootstrap[$path]) * 0x0100 + $bootstrap[$path][$str][1], SEEK_SET);
+          $bytes= fread($h, $bootstrap[$path][$str][0]);
+          if (FALSE === eval('?>'.$bytes)) {
+            xp::error('Bootstrap class loading failure at '.$str.' (file= '.$path.' / start= '.$bootstrap[$path][$str][1].', length= '.$bootstrap[$path][$str][0].')');
+          }
+          $type= 'BTSLDR';
+          break;
         }
       }
       
+      $t2= microtime();
+      list($tm1, $ts1)= explode(' ', $t1);
+      list($tm2, $ts2)= explode(' ', $t2);
+      $diff= ($ts2- $ts1) + ($tm2 - $tm1);
+      
+      // DEBUG
+      0 && printf("Class loading via %s took %0.3fs class= %s\n",
+        $type,
+        $diff,
+        $str
+      );
+
+      if (!class_exists(xp::reflect($str))) {
+        xp::error('Cannot include '.$str);
+      }
+            
       // Register class name and call static initializer if available
       xp::registry('class.'.$class, $str);
       is_callable(array($class, '__static')) && call_user_func(array($class, '__static'));
@@ -428,7 +498,6 @@
     ? getenv('SKELETON_PATH')
     : dirname(__FILE__).DIRECTORY_SEPARATOR
   ));
-  ini_set('include_path', SKELETON_PATH.PATH_SEPARATOR.ini_get('include_path'));
   define('LONG_MAX', is_int(2147483648) ? 9223372036854775807 : 2147483647);
   define('LONG_MIN', -LONG_MAX - 1);
 
Index: io/cca/ArchiveClassLoader.class.php
===================================================================
--- io/cca/ArchiveClassLoader.class.php	(revision 7907)
+++ io/cca/ArchiveClassLoader.class.php	(working copy)
@@ -55,43 +55,7 @@
      * @return  string
      */
     function loadClassBytes($name) {
-      $src= '';
-      $line= 0;
-      $tokens= token_get_all($this->archive->extract($name));
-      for ($i= 0, $s= sizeof($tokens); $i < $s; $i++) {
-        switch ($tokens[$i][0]) {
-          case T_FILE: 
-            $tokens[$i][1]= "'".strtr($name, '.', '/').'.class.php\''; 
-            break;
-            
-          case T_LINE:
-            $tokens[$i][1]= $line;
-            break;
-
-          case T_STRING:
-            if ('uses' == $tokens[$i][1] || 'implements' == $tokens[$i][1]) {
-              $o= $i+ 1;
-              while (')' != $tokens[$o][0]) {
-                if (T_CONSTANT_ENCAPSED_STRING == $tokens[$o][0]) {
-                  $used= trim($tokens[$o][1], '"\'');
-                  $this->archive->contains($used) && $this->loadClass($used);
-                }
-                $o++;
-              }
-            }
-            break;
-        }
-
-        if (is_array($tokens[$i])) {
-          $src.= $tokens[$i][1];
-          $line+= substr_count($tokens[$i][1], "\n");
-        } else {
-          $src.= $tokens[$i];
-          $line+= substr_count($tokens[$i], "\n");
-        }
-      }
-      
-      return $src;
+      return $this->archive->extract($name);
     }
     
     /**
@@ -128,5 +92,16 @@
       $c= &new XPClass($name);
       return $c;
     }
+    
+    /**
+     * Checks whether this loader can provide the requested class
+     *
+     * @access  public
+     * @param   string fqcn
+     * @return  bool
+     */
+    function providesClass($class) {
+      return $this->archive->contains($class);
+    }    
   }
 ?>
Index: io/cca/Archive.class.php
===================================================================
--- io/cca/Archive.class.php	(revision 7907)
+++ io/cca/Archive.class.php	(working copy)
@@ -97,6 +97,25 @@
     }
     
     /**
+     * Add a file by its bytes
+     *
+     * @access  public
+     * @param   string id the id under which this entry will be located
+     * @param   string path
+     * @param   string filename
+     * @param   string bytes
+     */
+    function addFileBytes($id, $path, $filename, $bytes) {
+      $this->_index[$id]= array(
+        $filename,
+        $path,
+        strlen($bytes),
+        -1,                 // Will be calculated by create()
+        $bytes
+      );
+    }
+    
+    /**
      * Create CCA archive
      *
      * @access  public
